tests/transaction_examples.rs

```diff
@@ -0,0 +1,81 @@
+#![allow(dead_code)]
+
+extern crate mongodb;
+
+use futures::Future;
+use mongodb::{
+    bson::{doc, Document},
+    error::{Result, TRANSIENT_TRANSACTION_ERROR, UNKNOWN_TRANSACTION_COMMIT_RESULT},
+    options::{Acknowledgment, ReadConcern, TransactionOptions, WriteConcern},
+    ClientSession,
+};
+
+async fn execute_transaction_with_retry<F, G>(
+    execute_transaction: F,
+    session: &mut ClientSession,
+) -> Result<()>
+where
+    F: Fn(&mut ClientSession) -> G,
+    G: Future<Output = Result<()>>,
+{
+    while let Err(err) = execute_transaction(session).await {
+        println!("Transaction aborted. Error returned during transaction.");
+        if err.contains_label(TRANSIENT_TRANSACTION_ERROR) {
+            println!("Encountered TransientTransactionError, retrying transaction.");
+            continue;
+        } else {
+            return Err(err);
+        }
+    }
+    Ok(())
+}
+
+async fn commit_with_retry(session: &mut ClientSession) -> Result<()> {
+    while let Err(err) = session.commit_transaction().await {
+        if err.contains_label(UNKNOWN_TRANSACTION_COMMIT_RESULT) {
+            println!("Encountered UnknownTransactionCommitResult, retrying commit operation.");
+            continue;
+        } else {
+            println!("Encountered non-retryable error during commit.");
+            return Err(err);
+        }
+    }
+    println!("Transaction committed.");
+    Ok(())
+}
+
+async fn execute_employee_info_transaction(session: &mut ClientSession) -> Result<()> {
+    let client = session.client();
+    let employees = client.database("hr").collection::<Document>("employees");
+    let events = client
+        .database("reporting")
+        .collection::<Document>("events");
+
+    employees
+        .update_one_with_session(
+            doc! { "employee": 3 },
+            doc! { "$set": { "status": "Inactive" } },
+            None,
+            session,
+        )
+        .await?;
+    events
+        .insert_one_with_session(
+            doc! { "employee": 3, "status": { "new": "Inactive", "old": "Active" } },
+            None,
+            session,
+        )
+        .await?;
+
+    commit_with_retry(session).await
+}
+
+async fn update_employee_info(session: &mut ClientSession) -> Result<()> {
+    let transaction_options = TransactionOptions::builder()
+        .read_concern(ReadConcern::snapshot())
+        .write_concern(WriteConcern::builder().w(Acknowledgment::Majority).build())
+        .build();
+    session.start_transaction(transaction_options).await?;
+
+    execute_transaction_with_retry(execute_employee_info_transaction, session).await
```

Comment:
Running into some lifetime issues here -- the compiler error indicates that the `&mut ClientSession` argument in the function pointer needs a lifetime, but adding a lifetime parameter causes a multiple mutable references issue in the loop on line 21. This issue doesn't appear in a similar loop in the docs examples in `src/client/session/mod.rs`, so it seems like it has something to do with the lifetime that I can't quite figure out.

Reply:
Got it. That macro isn't too daunting but IMO isn't ideal for basic transactions starter code to provide for users -- I think it might be best to deviate a bit from the other examples and eliminate the closure (e.g. by calling `execute_employee_info_transaction`) directly. Thoughts?

Reply:
Yeah I think you're right. Maybe we ought to just include the retry part directly in `update_employee_info` then instead of invoking the macro.